File: openmole/core/org.openmole.core.implementation/src/main/java/org/openmole/core/implementation/execution/local/LocalExecuter.java
/*
 *  Copyright (C) 2010 reuillon
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.openmole.core.implementation.execution.local;

import java.util.logging.Level;
import java.util.logging.Logger;

import org.openmole.core.model.execution.ExecutionState;
import org.openmole.core.model.execution.batch.SampleType;
import org.openmole.core.model.job.IJob;
import org.openmole.core.model.job.IMoleJob;
import org.openmole.core.model.task.IMoleTask;
import org.openmole.commons.tools.structure.Trio;
import org.openmole.core.model.job.State;

public class LocalExecuter implements Runnable {

    private static final Logger LOGGER = Logger.getLogger(LocalExecuter.class.getName());
    LocalExecutionEnvironment localEnvironment;
    boolean stop = false;

    public LocalExecuter(LocalExecutionEnvironment localEnvironment) {
        super();
        this.localEnvironment = localEnvironment;
    }

    @Override
    public void run() {

        while (!stop) {
            try {
                LocalExecutionJob executionJob = localEnvironment.takeNextjob();
                IJob job = executionJob.getJob();

                try {
                    executionJob.setState(ExecutionState.RUNNING);
                    Long running = System.currentTimeMillis();
                    executionJob.getEnvironment().sample(SampleType.WAITING, running - executionJob.getCreationTime(), job);

                    for (IMoleJob moleJob : job.getMoleJobs()) {
                        if (moleJob.getState() != State.CANCELED) {
                            LOGGER.log(Level.FINER, "New job group taken for execution: {0}", moleJob);

                            if (IMoleTask.class.isAssignableFrom(moleJob.getTask().getClass())) {
                                jobGoneIdle();
                            }
                            moleJob.perform();
                            moleJob.finished(moleJob.getContext());
                            LOGGER.log(Level.FINER, "End of job group execution: {0}", moleJob);
                        }
                    }
                    executionJob.setState(ExecutionState.DONE);
                    executionJob.getEnvironment().sample(SampleType.RUNNING, System.currentTimeMillis() - running, job);
                } finally {
                    localEnvironment.getJobRegistry().removeJob(job);
                }
            } catch (InterruptedException e) {
                if (!stop) {
                    LOGGER.log(Level.WARNING, "Interrupted despite stop is false.", e);
                }
            } catch (Throwable e) {
                LOGGER.log(Level.SEVERE, null, e);
            }
        }

    }

    protected boolean isStop() {
        return stop;
    }

    protected void setStop(boolean stop) {
        this.stop = stop;
    }

    public void jobGoneIdle() {
        localEnvironment.addExecuters(1);
        setStop(true);
    }
}

File: openmole/core/org.openmole.core.implementation/src/main/scala/org/openmole/core/implementation/execution/local/LocalExecuter.scala
/*
 * Copyright (C) 2010 reuillon
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.openmole.core.implementation.execution.local

import org.openmole.core.implementation.execution.StatisticRegistry
import org.openmole.core.implementation.execution.StatisticSample
import org.openmole.core.model.execution.ExecutionState
import org.openmole.core.model.execution.IExecutionJob
import org.openmole.core.model.job.IMoleJob
import org.openmole.core.model.job.State
import org.openmole.core.model.task.IMoleTask
import org.openmole.misc.eventdispatcher.EventDispatcher
import org.openmole.misc.tools.service.Logger
import scala.collection.JavaConversions._

object LocalExecuter extends Logger

class LocalExecuter(environment: LocalExecutionEnvironment) extends Runnable {

  import LocalExecuter._
  
  var stop: Boolean = false;

  override def run = {
    
    while (!stop) {
      val executionJob = environment.takeNextjob
      try {
        val job = executionJob.job
        executionJob.state = ExecutionState.RUNNING
        val running = System.currentTimeMillis

        for (moleJob <- job.moleJobs) {
          if (moleJob.state != State.CANCELED) {
            if (classOf[IMoleTask].isAssignableFrom(moleJob.task.getClass)) jobGoneIdle
            moleJob.perform
            
            moleJob.exception match {
              case None =>
              case Some(e) => 
                EventDispatcher.trigger(moleJob, new IMoleJob.ExceptionRaised(e, SEVERE))
                //EventDispatcher.trigger(executionJob, new IExecutionJob.ExceptionRaised(e, SEVERE))
                logger.log(SEVERE, "Error in user job execution, job state is FAILED.", e)
            }
          }
        }
        executionJob.state = ExecutionState.DONE
        StatisticRegistry.sample(environment, job, new StatisticSample(executionJob.timeStamps.head.time, running, System.currentTimeMillis))
      } catch {
        case (e: InterruptedException) => if (!stop) logger.log(WARNING, "Interrupted despite stop is false.", e)  
        case e => logger.log(SEVERE, null, e)
      } finally executionJob.state = ExecutionState.KILLED
    }
  }

  def jobGoneIdle {
    environment.addExecuters(1)
    stop = true
  }
}
